File: project/app/javascript/channels/group_chat_channel.js
import consumer from "channels/consumer";

export function ConnectGroupChatChannel(
  message_collection,
  room_id,
  room_members
) {
  return consumer.subscriptions.create(
    {
      channel: "GroupChatChannel",
      room_id: room_id,
    },
    {
      connected() {
        // Called when the subscription is ready for use on the server
      },

      disconnected() {
        this.unsubscribe();
        // Called when the subscription has been terminated by the server
      },

      received(message) {
        const member = room_members.get(message.user_id);

        switch (message.type) {
          case "msg":
            message_collection.add(message);
            message_collection.trigger("scroll");
            break;
          case "join":
            room_members.add(message.user);
            break;
          case "restore":
            member.set("position", "member");
            member.trigger("restore", member);
            break;
          case "position":
            member.set("position", message.position);
            break;
          case "mute":
            member.set("mute", message.mute);
            break;
        }
      },

      speak(current_user_message) {
        current_user_message["room_id"] = room_id;
        this.perform("speak", current_user_message);
      },
    }
  );
}

File: project/app/channels/group_chat_channel.rb
class GroupChatChannel < ApplicationCable::Channel
  def subscribed
    stream_from "group_chat_channel_#{params[:room_id]}"
  end

  def speak(msg)
    begin
      membership = GroupChatMembership.find_by_group_chat_room_id_and_user_id(msg['room_id'], msg['user_id'])
      raise ServiceError.new(:NotFound) if membership.nil?
      return if membership.mute?

      message = ChatMessage.create!(
        user_id: msg['user_id'],
        room_type: 'GroupChatRoom',
        room_id: msg['room_id'],
        message: msg['message']
      )
      msg['created_at'] = message.created_at
      ActionCable.server.broadcast("group_chat_channel_#{msg['room_id']}", msg)
    rescue => e
      put "[ERROR][GroupChatChannel_#{msg['room_id']}][speak] #{e.message}"
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end
end
